File: experiment-5/runs/language=go_model=claude-opus-4-8_tooling=none/rep1/internal/soccer/loader.go
// Context:
//   - This file reads the six provided CSV files and converts each into the
//     unified Match/Player model. Every source has its own column layout, date
//     format and quirks, so there is one loader per file plus shared helpers.
//   - Quirks handled here: a UTF-8 BOM on the FIFA header row; goal counts
//     written as floats ("1.0"); Brazilian DD/MM/YYYY dates; goals quoted as
//     strings; and the BR-Football file having no season column (season is
//     derived from the match date).
//   - Loaders are tolerant: a malformed individual row is skipped rather than
//     aborting the whole load, so a single bad line never takes down the server.
package soccer

import (
	"encoding/csv"
	"io"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

// The canonical file names expected inside the data directory.
const (
	fileBrasileirao  = "Brasileirao_Matches.csv"
	fileCup          = "Brazilian_Cup_Matches.csv"
	fileLibertadores = "Libertadores_Matches.csv"
	fileBRFootball   = "BR-Football-Dataset.csv"
	fileNovo         = "novo_campeonato_brasileiro.csv"
	fileFifa         = "fifa_data.csv"
)

var dateLayouts = []string{
	"2006-01-02 15:04:05",
	"2006-01-02",
	"02/01/2006",
	"2006.01.02",
}

// parseDate tries each known layout. Returns the time and whether it succeeded.
func parseDate(s string) (time.Time, bool) {
	s = strings.TrimSpace(s)
	if s == "" {
		return time.Time{}, false
	}
	for _, layout := range dateLayouts {
		if t, err := time.Parse(layout, s); err == nil {
			return t, true
		}
	}
	return time.Time{}, false
}

// parseGoals parses a goal count that may be written as "2", "2.0" or "".
func parseGoals(s string) (int, bool) {
	s = strings.TrimSpace(s)
	if s == "" {
		return 0, false
	}
	if n, err := strconv.Atoi(s); err == nil {
		return n, true
	}
	if f, err := strconv.ParseFloat(s, 64); err == nil {
		return int(f), true
	}
	return 0, false
}

func parseInt(s string) int {
	n, _ := strconv.Atoi(strings.TrimSpace(s))
	return n
}

// openCSV opens a file and returns a configured csv.Reader.
func openCSV(path string) (*os.File, *csv.Reader, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, nil, err
	}
	r := csv.NewReader(f)
	r.FieldsPerRecord = -1 // tolerate ragged rows
	r.LazyQuotes = true
	return f, r, nil
}

// headerIndex builds a name->column map, trimming a leading BOM if present.
func headerIndex(header []string) map[string]int {
	idx := make(map[string]int, len(header))
	for i, h := range header {
		h = strings.TrimPrefix(h, "\ufeff")
		idx[strings.TrimSpace(h)] = i
	}
	return idx
}

func get(row []string, idx map[string]int, name string) string {
	if i, ok := idx[name]; ok && i < len(row) {
		return strings.TrimSpace(row[i])
	}
	return ""
}

// finishMatch fills in the derived fields (clean names, normalized keys, season
// fallback) shared by every match loader.
func finishMatch(m *Match, source string) {
	m.HomeNorm = NormalizeTeam(m.HomeTeam)
	m.AwayNorm = NormalizeTeam(m.AwayTeam)
	m.HomeTeam = CleanTeamName(m.HomeTeam)
	m.AwayTeam = CleanTeamName(m.AwayTeam)
	if m.Season == 0 && m.HasDate {
		m.Season = m.Date.Year()
	}
	m.Sources = []string{source}
}

// loadBrasileirao reads Brasileirao_Matches.csv (Série A, 2012+).
func loadBrasileirao(path string) ([]Match, error) {
	return loadGeneric(path, fileBrasileirao, CompSerieA, func(row []string, idx map[string]int) (Match, bool) {
		m := Match{Competition: CompSerieA}
		m.HomeTeam = get(row, idx, "home_team")
		m.AwayTeam = get(row, idx, "away_team")
		m.HomeState = get(row, idx, "home_team_state")
		m.AwayState = get(row, idx, "away_team_state")
		m.Round = get(row, idx, "round")
		m.Season = parseInt(get(row, idx, "season"))
		if t, ok := parseDate(get(row, idx, "datetime")); ok {
			m.Date, m.HasDate = t, true
		}
		hg, ok1 := parseGoals(get(row, idx, "home_goal"))
		ag, ok2 := parseGoals(get(row, idx, "away_goal"))
		m.HomeGoals, m.AwayGoals, m.HasScore = hg, ag, ok1 && ok2
		return m, m.HomeTeam != "" && m.AwayTeam != ""
	})
}

// loadCup reads Brazilian_Cup_Matches.csv (Copa do Brasil).
func loadCup(path string) ([]Match, error) {
	return loadGeneric(path, fileCup, CompCopaDoBrasil, func(row []string, idx map[string]int) (Match, bool) {
		m := Match{Competition: CompCopaDoBrasil}
		m.HomeTeam = get(row, idx, "home_team")
		m.AwayTeam = get(row, idx, "away_team")
		m.Round = get(row, idx, "round")
		m.Season = parseInt(get(row, idx, "season"))
		if t, ok := parseDate(get(row, idx, "datetime")); ok {
			m.Date, m.HasDate = t, true
		}
		hg, ok1 := parseGoals(get(row, idx, "home_goal"))
		ag, ok2 := parseGoals(get(row, idx, "away_goal"))
		m.HomeGoals, m.AwayGoals, m.HasScore = hg, ag, ok1 && ok2
		return m, m.HomeTeam != "" && m.AwayTeam != ""
	})
}

// loadLibertadores reads Libertadores_Matches.csv (Copa Libertadores).
func loadLibertadores(path string) ([]Match, error) {
	return loadGeneric(path, fileLibertadores, CompLibertadores, func(row []string, idx map[string]int) (Match, bool) {
		m := Match{Competition: CompLibertadores}
		m.HomeTeam = get(row, idx, "home_team")
		m.AwayTeam = get(row, idx, "away_team")
		m.Stage = get(row, idx, "stage")
		m.Season = parseInt(get(row, idx, "season"))
		if t, ok := parseDate(get(row, idx, "datetime")); ok {
			m.Date, m.HasDate = t, true
		}
		hg, ok1 := parseGoals(get(row, idx, "home_goal"))
		ag, ok2 := parseGoals(get(row, idx, "away_goal"))
		m.HomeGoals, m.AwayGoals, m.HasScore = hg, ag, ok1 && ok2
		return m, m.HomeTeam != "" && m.AwayTeam != ""
	})
}

// loadBRFootball reads BR-Football-Dataset.csv (Série A/B/C and Copa do Brasil,
// with extended statistics). It has no season column, so season is the year of
// the match date.
func loadBRFootball(path string) ([]Match, error) {
	return loadGeneric(path, fileBRFootball, "", func(row []string, idx map[string]int) (Match, bool) {
		m := Match{Competition: canonicalCompetition(get(row, idx, "tournament"))}
		m.HomeTeam = get(row, idx, "home")
		m.AwayTeam = get(row, idx, "away")
		if t, ok := parseDate(get(row, idx, "date")); ok {
			m.Date, m.HasDate = t, true
		}
		hg, ok1 := parseGoals(get(row, idx, "home_goal"))
		ag, ok2 := parseGoals(get(row, idx, "away_goal"))
		m.HomeGoals, m.AwayGoals, m.HasScore = hg, ag, ok1 && ok2
		return m, m.HomeTeam != "" && m.AwayTeam != "" && m.Competition != ""
	})
}

// loadNovo reads novo_campeonato_brasileiro.csv (historical Série A 2003-2019).
func loadNovo(path string) ([]Match, error) {
	return loadGeneric(path, fileNovo, CompSerieA, func(row []string, idx map[string]int) (Match, bool) {
		m := Match{Competition: CompSerieA}
		m.HomeTeam = get(row, idx, "Equipe_mandante")
		m.AwayTeam = get(row, idx, "Equipe_visitante")
		m.HomeState = get(row, idx, "Mandante_UF")
		m.AwayState = get(row, idx, "Visitante_UF")
		m.Round = get(row, idx, "Rodada")
		m.Arena = get(row, idx, "Arena")
		m.Season = parseInt(get(row, idx, "Ano"))
		if t, ok := parseDate(get(row, idx, "Data")); ok {
			m.Date, m.HasDate = t, true
		}
		hg, ok1 := parseGoals(get(row, idx, "Gols_mandante"))
		ag, ok2 := parseGoals(get(row, idx, "Gols_visitante"))
		m.HomeGoals, m.AwayGoals, m.HasScore = hg, ag, ok1 && ok2
		return m, m.HomeTeam != "" && m.AwayTeam != ""
	})
}

// loadGeneric drives the shared read loop: read header, map columns, apply the
// per-file row converter, finish derived fields, skip rows the converter rejects.
func loadGeneric(path, source, _ string, convert func([]string, map[string]int) (Match, bool)) ([]Match, error) {
	f, r, err := openCSV(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	header, err := r.Read()
	if err != nil {
		return nil, err
	}
	idx := headerIndex(header)

	var out []Match
	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			continue // skip malformed line
		}
		m, ok := convert(row, idx)
		if !ok {
			continue
		}
		finishMatch(&m, source)
		out = append(out, m)
	}
	return out, nil
}

// loadPlayers reads fifa_data.csv into Player records.
func loadPlayers(path string) ([]Player, error) {
	f, r, err := openCSV(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	header, err := r.Read()
	if err != nil {
		return nil, err
	}
	idx := headerIndex(header)

	var out []Player
	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			continue
		}
		p := Player{
			ID:            parseInt(get(row, idx, "ID")),
			Name:          get(row, idx, "Name"),
			Age:           parseInt(get(row, idx, "Age")),
			Nationality:   get(row, idx, "Nationality"),
			Overall:       parseInt(get(row, idx, "Overall")),
			Potential:     parseInt(get(row, idx, "Potential")),
			Club:          get(row, idx, "Club"),
			Position:      get(row, idx, "Position"),
			JerseyNumber:  get(row, idx, "Jersey Number"),
			Height:        get(row, idx, "Height"),
			Weight:        get(row, idx, "Weight"),
			PreferredFoot: get(row, idx, "Preferred Foot"),
			Value:         get(row, idx, "Value"),
			Wage:          get(row, idx, "Wage"),
		}
		if p.Name == "" {
			continue
		}
		p.NameNorm = normalizeText(p.Name)
		p.ClubNorm = normalizeText(p.Club)
		out = append(out, p)
	}
	return out, nil
}

// loadAllMatches loads every match file present in dir. Missing files are
// skipped (so the loader works even if a subset of data is available); a parse
// error on a present file is returned.
func loadAllMatches(dir string) ([]Match, error) {
	type fileLoader struct {
		name string
		fn   func(string) ([]Match, error)
	}
	loaders := []fileLoader{
		{fileBrasileirao, loadBrasileirao},
		{fileCup, loadCup},
		{fileLibertadores, loadLibertadores},
		{fileBRFootball, loadBRFootball},
		{fileNovo, loadNovo},
	}
	var all []Match
	for _, l := range loaders {
		path := filepath.Join(dir, l.name)
		if _, err := os.Stat(path); err != nil {
			continue // file not present
		}
		ms, err := l.fn(path)
		if err != nil {
			return nil, err
		}
		all = append(all, ms...)
	}
	return all, nil
}

File: experiment-5/runs/language=rust_model=claude-opus-4-8_tooling=none/rep3/src/loader.rs
//! ============================================================================
//! Module: loader
//! Project: Brazilian Soccer MCP Server (Rust)
//!
//! Context:
//!   Reads the six provided CSV files from `data/kaggle/` and converts each row
//!   into the normalized `Match` / `Player` domain types. Each source file has
//!   its own quirks that are handled here:
//!     - Brasileirao_Matches.csv  : team names carry "-UF" suffix, int goals.
//!     - Brazilian_Cup_Matches.csv: long club names, no state column.
//!     - Libertadores_Matches.csv : goals are quoted strings, has `stage`.
//!     - BR-Football-Dataset.csv  : goals are floats ("1.0"), `tournament`
//!                                  column distinguishes Serie A/B/C & Cup.
//!     - novo_campeonato_brasileiro.csv : Portuguese headers, DD/MM/YYYY dates.
//!     - fifa_data.csv            : 89-column player table; we keep a subset.
//!
//!   Parsing is deliberately lenient: a row that fails to parse a numeric field
//!   is skipped rather than aborting the whole load, so a few dirty rows never
//!   take down the server. Goal fields are parsed via `parse_goal` which copes
//!   with ints, floats and quoted values.
//! ============================================================================

use crate::model::{Competition, Match, Player};
use crate::normalize::{normalize_date, normalize_team};
use std::error::Error;
use std::path::Path;

/// Parse a goal count that may be an int ("2"), float ("2.0") or empty.
fn parse_goal(s: &str) -> Option<i32> {
    let t = s.trim();
    if t.is_empty() {
        return None;
    }
    if let Ok(i) = t.parse::<i32>() {
        return Some(i);
    }
    t.parse::<f64>().ok().map(|f| f.round() as i32)
}

fn parse_season(s: &str) -> Option<i32> {
    s.trim().parse::<i32>().ok()
}

/// Build a CSV reader for a path with flexible field counts.
fn reader(path: &Path) -> Result<csv::Reader<std::fs::File>, Box<dyn Error>> {
    let rdr = csv::ReaderBuilder::new()
        .has_headers(true)
        .flexible(true)
        .from_path(path)?;
    Ok(rdr)
}

/// Map header names to indices for robust, order-independent access.
fn header_index(headers: &csv::StringRecord) -> std::collections::HashMap<String, usize> {
    headers
        .iter()
        .enumerate()
        .map(|(i, h)| (h.trim().trim_start_matches('\u{feff}').to_lowercase(), i))
        .collect()
}

fn get<'a>(rec: &'a csv::StringRecord, idx: &std::collections::HashMap<String, usize>, key: &str) -> &'a str {
    idx.get(key).and_then(|&i| rec.get(i)).unwrap_or("").trim()
}

/// Load the Brasileirão Serie A matches file.
pub fn load_brasileirao(path: &Path) -> Result<Vec<Match>, Box<dyn Error>> {
    let mut rdr = reader(path)?;
    let idx = header_index(rdr.headers()?);
    let mut out = Vec::new();
    for rec in rdr.records() {
        let rec = rec?;
        let (h, a) = (get(&rec, &idx, "home_team"), get(&rec, &idx, "away_team"));
        let (hg, ag) = (
            parse_goal(get(&rec, &idx, "home_goal")),
            parse_goal(get(&rec, &idx, "away_goal")),
        );
        if h.is_empty() || a.is_empty() {
            continue;
        }
        let (hg, ag) = match (hg, ag) {
            (Some(x), Some(y)) => (x, y),
            _ => continue,
        };
        let date = normalize_date(get(&rec, &idx, "datetime"));
        let round = get(&rec, &idx, "round");
        out.push(Match {
            competition: Competition::Brasileirao,
            date,
            home_team: normalize_team(h),
            away_team: normalize_team(a),
            home_team_raw: h.to_string(),
            away_team_raw: a.to_string(),
            home_goal: hg,
            away_goal: ag,
            season: parse_season(get(&rec, &idx, "season")),
            round: if round.is_empty() { None } else { Some(round.to_string()) },
            source: "Brasileirao_Matches.csv",
        });
    }
    Ok(out)
}

/// Load the Copa do Brasil matches file.
pub fn load_cup(path: &Path) -> Result<Vec<Match>, Box<dyn Error>> {
    let mut rdr = reader(path)?;
    let idx = header_index(rdr.headers()?);
    let mut out = Vec::new();
    for rec in rdr.records() {
        let rec = rec?;
        let (h, a) = (get(&rec, &idx, "home_team"), get(&rec, &idx, "away_team"));
        let (hg, ag) = match (
            parse_goal(get(&rec, &idx, "home_goal")),
            parse_goal(get(&rec, &idx, "away_goal")),
        ) {
            (Some(x), Some(y)) => (x, y),
            _ => continue,
        };
        if h.is_empty() || a.is_empty() {
            continue;
        }
        let round = get(&rec, &idx, "round");
        out.push(Match {
            competition: Competition::CopaDoBrasil,
            date: normalize_date(get(&rec, &idx, "datetime")),
            home_team: normalize_team(h),
            away_team: normalize_team(a),
            home_team_raw: h.to_string(),
            away_team_raw: a.to_string(),
            home_goal: hg,
            away_goal: ag,
            season: parse_season(get(&rec, &idx, "season")),
            round: if round.is_empty() { None } else { Some(round.to_string()) },
            source: "Brazilian_Cup_Matches.csv",
        });
    }
    Ok(out)
}

/// Load the Copa Libertadores matches file (goals quoted, has `stage`).
pub fn load_libertadores(path: &Path) -> Result<Vec<Match>, Box<dyn Error>> {
    let mut rdr = reader(path)?;
    let idx = header_index(rdr.headers()?);
    let mut out = Vec::new();
    for rec in rdr.records() {
        let rec = rec?;
        let (h, a) = (get(&rec, &idx, "home_team"), get(&rec, &idx, "away_team"));
        let (hg, ag) = match (
            parse_goal(get(&rec, &idx, "home_goal")),
            parse_goal(get(&rec, &idx, "away_goal")),
        ) {
            (Some(x), Some(y)) => (x, y),
            _ => continue,
        };
        if h.is_empty() || a.is_empty() {
            continue;
        }
        let stage = get(&rec, &idx, "stage");
        out.push(Match {
            competition: Competition::Libertadores,
            date: normalize_date(get(&rec, &idx, "datetime")),
            home_team: normalize_team(h),
            away_team: normalize_team(a),
            home_team_raw: h.to_string(),
            away_team_raw: a.to_string(),
            home_goal: hg,
            away_goal: ag,
            season: parse_season(get(&rec, &idx, "season")),
            round: if stage.is_empty() { None } else { Some(stage.to_string()) },
            source: "Libertadores_Matches.csv",
        });
    }
    Ok(out)
}

/// Load the extended BR-Football statistics file (float goals, `tournament`).
pub fn load_br_football(path: &Path) -> Result<Vec<Match>, Box<dyn Error>> {
    let mut rdr = reader(path)?;
    let idx = header_index(rdr.headers()?);
    let mut out = Vec::new();
    for rec in rdr.records() {
        let rec = rec?;
        let (h, a) = (get(&rec, &idx, "home"), get(&rec, &idx, "away"));
        let (hg, ag) = match (
            parse_goal(get(&rec, &idx, "home_goal")),
            parse_goal(get(&rec, &idx, "away_goal")),
        ) {
            (Some(x), Some(y)) => (x, y),
            _ => continue,
        };
        if h.is_empty() || a.is_empty() {
            continue;
        }
        let tournament = get(&rec, &idx, "tournament");
        let date = normalize_date(get(&rec, &idx, "date"));
        out.push(Match {
            competition: Competition::from_text(tournament),
            season: date.split('-').next().and_then(|y| y.parse().ok()),
            date,
            home_team: normalize_team(h),
            away_team: normalize_team(a),
            home_team_raw: h.to_string(),
            away_team_raw: a.to_string(),
            home_goal: hg,
            away_goal: ag,
            round: None,
            source: "BR-Football-Dataset.csv",
        });
    }
    Ok(out)
}

/// Load the historical Brasileirão 2003-2019 file (Portuguese headers).
pub fn load_novo(path: &Path) -> Result<Vec<Match>, Box<dyn Error>> {
    let mut rdr = reader(path)?;
    let idx = header_index(rdr.headers()?);
    let mut out = Vec::new();
    for rec in rdr.records() {
        let rec = rec?;
        let h = get(&rec, &idx, "equipe_mandante");
        let a = get(&rec, &idx, "equipe_visitante");
        let (hg, ag) = match (
            parse_goal(get(&rec, &idx, "gols_mandante")),
            parse_goal(get(&rec, &idx, "gols_visitante")),
        ) {
            (Some(x), Some(y)) => (x, y),
            _ => continue,
        };
        if h.is_empty() || a.is_empty() {
            continue;
        }
        let round = get(&rec, &idx, "rodada");
        out.push(Match {
            competition: Competition::Brasileirao,
            date: normalize_date(get(&rec, &idx, "data")),
            home_team: normalize_team(h),
            away_team: normalize_team(a),
            home_team_raw: h.to_string(),
            away_team_raw: a.to_string(),
            home_goal: hg,
            away_goal: ag,
            season: parse_season(get(&rec, &idx, "ano")),
            round: if round.is_empty() { None } else { Some(round.to_string()) },
            source: "novo_campeonato_brasileiro.csv",
        });
    }
    Ok(out)
}

/// Load the FIFA player database (keep a useful subset of columns).
pub fn load_players(path: &Path) -> Result<Vec<Player>, Box<dyn Error>> {
    let mut rdr = reader(path)?;
    let idx = header_index(rdr.headers()?);
    let mut out = Vec::new();
    for rec in rdr.records() {
        let rec = rec?;
        let name = get(&rec, &idx, "name");
        if name.is_empty() {
            continue;
        }
        let overall = get(&rec, &idx, "overall").parse::<i32>().unwrap_or(0);
        let potential = get(&rec, &idx, "potential").parse::<i32>().unwrap_or(overall);
        out.push(Player {
            id: get(&rec, &idx, "id").parse::<i64>().unwrap_or(0),
            name: name.to_string(),
            age: get(&rec, &idx, "age").parse::<i32>().ok(),
            nationality: get(&rec, &idx, "nationality").to_string(),
            overall,
            potential,
            club: get(&rec, &idx, "club").to_string(),
            position: get(&rec, &idx, "position").to_string(),
            jersey_number: get(&rec, &idx, "jersey number").parse::<i32>().ok(),
            height: get(&rec, &idx, "height").to_string(),
            weight: get(&rec, &idx, "weight").to_string(),
        });
    }
    Ok(out)
}
